### PR TITLE
Mod 4

### PR DESCRIPTION
**Pull request by Awlter**

### CODE REVIEW

File: app/views/relationships/index.html.haml [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+    %tbody
+      - @leaders.each do |leader|
+        %tr
+          %td
+            %img(src="http://www.gravatar.com/avatar/#{Digest::MD5.hexdigest(leader.email.downcase)}?s=40")
+            = link_to leader.full_name, leader
+          %td.extra-padding= leader.queue_items.count
+          %td.extra-padding= leader.followers.count
+          %td.extra-padding
+            - relationship = Relationship.find_by(leader: leader, follower: current_user)
```
Comment (Reltre) [inline app/views/relationships/index.html.haml]: I don't recommend using business logic like this in your views, though I realize that this is only one line. I recommend trying something a bit more compact. For instance, you could do something like  
`= link_to Relationship.where(leader: leader, follower: current_user)`
or using find_by directly as that first argument.
You could also make a helper method in under relationships helpers folder.

File: spec/controllers/relationships_controller_spec.rb [hunk 2]
```diff
@@ -0,0 +1,89 @@
+require 'spec_helper'
+
+Rspec.describe RelationshipsController do
```
Comment (Reltre) [inline spec/controllers/relationships_controller_spec.rb]: You may leave off the initial `Rspec` here. Though, it doesn't hurt to be explicit in your method calls.

File: spec/controllers/relationships_controller_spec.rb [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+        user = Fabricate(:user)
+        set_current_user(user)
+        other_user = Fabricate(:user)
+        relationship = Fabricate(:relationship, follower: user, leader: other_user)
+        get :index
+        expect(assigns(:leaders)).to eq [other_user]
+      end
+    end
+
+    context "as a guest" do
```
Comment (Reltre) [inline spec/controllers/relationships_controller_spec.rb:16]: I think this is a nice touch. Hmm, though keep in mind that at the moment you require that a user be logged in for all relationship actions. With that in mind, I would add the call to `it_behaves_like "requires sign in"` to the `create` example group as well

File: spec/support/macros.rb [hunk 4]
```diff
@@ -8,4 +8,8 @@ def sign_in(user=nil)
   fill_in "Email Address", with: user.email
   fill_in "Password", with: user.password
   click_button "Sign in"
+end
+
+def click_video(video)
```
Comment (Reltre) [inline spec/support/macros.rb:13]: You don't have to do this right away, but I recommend that you eventually create separate macro files, one for feature specs and one for controller specs.

File: spec/features/following_users_spec.rb [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+    user2 = Fabricate(:user)
+    review = Fabricate(:review, video: video, user: user2)
+    sign_in(user1)
+
+    click_video(video)
+    click_link(user2.full_name)
+    expect(page).to have_content "#{user2.full_name}'s video collections"
+
+    click_link "Follow"
+    expect(page).to have_content "People I Follow"
```
Comment (Reltre) [inline spec/features/following_users_spec.rb]: Consider moving expectations like this to custom made expectation methods that you create yourself. You can define these at the bottom of a feature spec.

File: config/routes.rb [hunk 6]
```diff
@@ -7,13 +7,16 @@
   get '/sign_out', to: 'sessions#destroy'
   get '/home', to: 'videos#index'
 
+  resources :users, only: [:create, :show]
+  resources :categorys, only: [:show]
```
Comment (Reltre) [inline config/routes.rb]: `category` pluralizes to `categories`

File: db/migrate/20171106074703_add_random_token_to_user.rb [hunk 7]
```diff
@@ -0,0 +1,9 @@
+class AddRandomTokenToUser < ActiveRecord::Migration
+  def change
+    add_column :users, :token, :string
+    User.all.each do |user|
```
Comment (Reltre) [inline db/migrate/20171106074703_add_random_token_to_user.rb]: I don't recommend altering data in your db migrations. Database migrations should only be to set up your schema. For any data related changes, add that to your seed file or a separate rake task. 
Unless you plan to have a few users when someone first runs your application, then I would leave this data update out entirely. Probably best to run this from the console.

File: spec/controllers/users_controller_spec.rb [hunk 8]
```diff
@@ -38,5 +59,31 @@
...
+    context "sending email" do
+      context "with valid input" do
+        after do
+          ActionMailer::Base.deliveries.clear
+        end
+
+        it "sends email" do
+          user_attributes = Fabricate.attributes_for(:user)
+          post :create, user: user_attributes
+          expect(ActionMailer::Base.deliveries.last.to).to eq [user_attributes[:email]]
```
Comment (Reltre) [inline spec/controllers/users_controller_spec.rb:72]: This is ok, though ideally you'll want to try and only have a single expectation per test case in your controller specs. 

I recommend also testing for the correct subject as well.

**Comment by Awlter** [comment]
Thanks a lot Tyler!

**Comment by Reltre** [comment]
yw, glad to help 👍

**Pull request merged**